File: run/dataUseEx.m
% Tyson Reimer
% University of Manitoba
% October 16th, 2019

clear all;
close all;
clc;

%%  Define variables specifying which file to load

% The generation of data to use, must be in ['one', 'two']
gen = 'one';

% The type of data to use, must be in ['iczt', 'idft', 'fd']
% NOTE: 'iczt' is the type that is expected for the plot below
dataType = 'iczt';

% The type of sparams to use, must be in ['s11', 's21']
sparam = 's11';

% The reference calibration type to use, must be in ['emp', 'adi']
calType = 'emp';

%% Load the scan data to an array variable

% Get the file name to load
fileName = sprintf('%s_data_%s_%s.mat', dataType, sparam, calType);

% Get the path to the file's parent directory
fileDir = sprintf('datasets/gen-%s/clean/', gen);

% Get the path to the file 
filePath = fullfile(fileDir, fileName);

scanData = load(filePath);  % Load the data

% Extract the data array from the struct dtype it is stored in
scanDataFieldName = fieldnames(scanData);
scanData = getfield(scanData, scanDataFieldName{1});

%% Load the metadata to a variable

% Define the file name and full path to the metadata file
fileName = sprintf('md_list_%s_%s.mat', sparam, calType);
filePath = fullfile(fileDir, fileName);

metadata = load(filePath);  % Load the metadata

% Extract the metadata array from the struct dytpe it is stored in
metadataFieldName = fieldnames(metadata);
metadata = getfield(metadata, metadataFieldName{1});

%% Plot a sample sinogram - USE 'iczt' AS dataType FOR THIS

sampleIdx = 1;  % The index for the sample that will be plotted

% Take the abs-value and reshape for imshow
dataToPlot = abs(scanData(sampleIdx, :, :));
dataToPlot = reshape(dataToPlot, [size(dataToPlot, 2), ...
                                    size(dataToPlot, 3)]);

% Make the figure
figure;
imagesc(dataToPlot, 'y', linspace(0, 6, 1024));
title('Sample Sinogram');
xlabel('Antenna Position');
ylabel('Time of Response (ns)');
yticks(round(linspace(0, 6, 10), 3));
colorbar;

%% Print metadata content

% Get the metadata for this scan, extract it from struct dtype
thisMetadata = metadata{sampleIdx};
mdFieldNames = fieldnames(thisMetadata);

% For each info piece stored in the metadata
for mdIdx = 1 : length(mdFieldNames)
    
    % Get the value for this info piece for the sample scan
    infoVal = getfield(thisMetadata, mdFieldNames{mdIdx});
    infoStr = sprintf('%20s', mdFieldNames{mdIdx});
    
    if ~ isnan(infoVal)  % If the value was not NaN
        
        % Define the string that will be printed
        strToPrint = strcat(infoStr, '\t', string(infoVal));
        
    else  % If the value was NaN
        strToPrint = strcat(infoStr, '\t', 'NaN');
    end
    fprintf(strcat(strToPrint, '\n'));
end

File: run/data_use_ex.py
"""
Tyson Reimer
University of Manitoba
October 16th, 2019
"""

import os
import numpy as np
import matplotlib.pyplot as plt

from umbmid import get_proj_path, verify_path, get_script_logger
from umbmid.loadsave import load_pickle
from umbmid.content import report_metadata_content

###############################################################################

__DATA_DIR = os.path.join(get_proj_path(), 'datasets')
__OUTPUT_DIR = os.path.join(get_proj_path(), 'output/figs/')
verify_path(__OUTPUT_DIR)

###############################################################################

# Define the generation of data
gen = 'one'

# Define the type of data, must be in ['iczt', 'idft', 'fd']
# NOTE: for plotting the sinogram, 'iczt' is expected
data_type = 'iczt'

# Define the type of s-params to plot, must be in ['s11', 's21']
sparam = 's11'

# Define the type of reference calibration performed, must
# be in ['emp', 'adi']
cal_type = 'emp'

###############################################################################

# Define the data directory for this generation
this_data_dir = os.path.join(__DATA_DIR, 'gen-%s/clean/' % gen)

# Load the scan data
scan_data = load_pickle(os.path.join(this_data_dir, '%s_data_%s_%s.pickle'
                                     % (data_type, sparam, cal_type)))

###############################################################################

# Load the metadata
metadata = load_pickle(os.path.join(this_data_dir, 'md_list_%s_%s.pickle'
                                    % (sparam, cal_type)))

logger = get_script_logger(__file__)

report_metadata_content(metadata, logger)

###############################################################################

# The index for the scan that will be plotted
sample_idx = 0

# Take the abs-value and get the data that will be plotted
data_to_plot = np.abs(scan_data[0, :, :])

# Plot the data
plt.figure()
plt.imshow(data_to_plot, cmap='inferno', aspect='auto')
plt.title('Sample Sinogram')
plt.xlabel('Antenna Position')
plt.ylabel('Time of Response (ns)')
plt.yticks(np.round(np.linspace(0, 1024, 10)),
           ['%.2f' % ii for ii in np.linspace(0, 6, 10)])
plt.colorbar()
plt.show()

# Save the figure, can comment-out to not-save
plt.savefig(os.path.join(__OUTPUT_DIR, 'PythonExampleSinogram.png'),
            dpi=300)

###############################################################################

# Define the metadata for this experiment
this_md = metadata[sample_idx]

for info_key in this_md.keys():  # For each piece of info in metadata

    # Print the value for this piece of metadata
    print('%20s:\t%s' % (info_key, this_md[info_key]))
